File: dutch_national_flag/java/TestThreeWayPartition.java
import java.util.Random;

public class TestThreeWayPartition
{
	public static void testPartition()
	{
		Random myrand = new Random();
		final int size = 100;
		int[] array = new int[size];
		int[] cnt = { 0, 0, 0 };
		int num = 0;
		for(int i = 0; i < size; i++)
		{
			num = myrand.nextInt(3);
			array[i] = num;
			cnt[num] += 1;
		}
		ThreeWayPartition.partition(array);
		for(int i = 0; i < size; i++)
		{
			if(array[i] != 0 && array[i - 1] > array[i])
				System.out.println("Partition Error: order");
			cnt[array[i]]--;
		}
		if(!(array[0] == array[1] && array[1] == array[2] && array[2] == 0))
			System.out.println("Partition Error: swap");
	}
	public static void main(String[] args)
	{
		for(int i = 0; i < 100; i++)
			testPartition();
	}
}

File: dutch_national_flag/c/test_three_way_partition.c
#include "three_way_partition.h"

#include <assert.h>
#include <stdlib.h>
#include <time.h>

void test_three_way_partition()
{
	const int size = 1000;
	int array[size];
	int cnt[3] = { 0, 0, 0 };
	int num = 0;
	for(int i = 0; i < size; i++)
	{
		num = rand() % 3;
		array[i] = num;
		cnt[num] += 1;
	}
	three_way_partition(array, size, default_is_head, default_is_tail);
	for(int i = 0; i < size; i++)
	{
		assert(i == 0 || array[i - 1] <= array[i]);
		cnt[array[i]]--;
	}
	assert(array[0] == array[1] && array[1] == array[2] && array[2] == 0);
}

int main()
{
	srand(time(0));
	test_three_way_partition();
	return 0;
}
